tensorflow/stream_executor/dnn.h

```diff
@@ -1301,6 +1413,17 @@ class DnnSupport {
       AlgorithmDesc algorithm_desc, DeviceMemory<uint8> scratch_memory,
       ProfileResult* output_profile_result) = 0;
 
+  virtual port::Status DoConvolve(
```

Comment:
Since these are CUDA-specific APIs, can you move it to `CudnnDnnSupport` and make them non-virtual? It's unlikely to be overridden by other subclasses. Users would do `dynamic_cast<CudnnDnnSupport*>` and then use it.

Reply:
Let me make sure that I understand you correctly. The current code structure is like:
```cpp
conv_ops.h:
stream->ConvolveWithExecutionPlan();

stream.h:
port::Status ConvolveWithExecutionPlan(){
  dnn::DnnSupport *dnn;
  dnn->DoConvolve();
}

dnn.h:
class DnnSupport {
  virtual port::Status DoConvolve();
} 

cuda_dnn.h:
class CudnnSupport : public dnn::DnnSupport {
  port::Status DoConvolve() override;
}
```
After your proposed move, the code will be like:
```cpp
conv_ops.h:
stream->ConvolveWithExecutionPlan();

stream.h:
port::Status ConvolveWithExecutionPlan(){
  dnn::DnnSupport *dnn = parent_->AsDnn();  
  gpu::CudnnSupport *cudnn_dnn = dynamic_cast<gpu::CudnnSupport *>(dnn);

  dnn->DoConvolveWithExecutionPlan();
}

dnn.h:
class DnnSupport {
  virtual port::Status DoConvolve();
} 

cuda_dnn.h:
class CudnnSupport : public dnn::DnnSupport {
  port::Status DoConvolveWithExecutionPlan();
}
```
Is that correct?

Reply:
That's correct. Thanks!

Reply:
Done.